### PR TITLE
Make pragmas look clickable in docs

### PR DESCRIPTION
**Pull request by Quelklef**
Hopefully I found the right file!

In documentation, the clickable `{..}` pragmas currently don't look clickable. They look like normal code and, given that it's not unreasonable to think that `{..}` is valid (if superfluous) code, that's what some users might think. I sure did.

### CODE REVIEW

**Comment by Araq** [comment]
Sorry, can you give us some nice before-after screenshots? :-)

**Comment by Quelklef** [comment]
It's nothing too wild...

Pragmas hidden
Pragmas shown

**Comment by Quelklef** [comment]
A small alternative is to have a line between the begin and end pragma tags:

image

image

But it doesn't look as clean when the pragmas are shown. I could try to make the open pragma also have a border when the pragmas are shown, but I'm not sure how the clicking quite works yet.

**Comment by dom96** [comment]
Wouldn't it make more sense to add a dotted line beneath these?

**Comment by Quelklef** [comment]
That could also work:

screenshot from 2018-07-02 15-07-13
screenshot from 2018-07-02 15-07-40

**Comment by Araq** [comment]
The line between the dots is better, but still too subtle.

**Comment by dom96** [comment]
Maybe something like this?

screen shot 2018-07-03 at 12 52 28

Alternatives:

screen shot 2018-07-03 at 12 53 10

screen shot 2018-07-03 at 12 52 52

**Comment by Quelklef** [comment]
Or perhaps something like:
screenshot from 2018-07-03 07-59-38
screenshot from 2018-07-03 08-01-41
It's terser, but looks a little strange with `...` between two `.`s.

A possible resolution is (with the open graphics unchanged):
screenshot from 2018-07-03 08-03-37
but then closed pragmas pragmas aren't `{.` and `.}`...

**Comment by dom96** [comment]
>> 
>
> Yeah, I like this solution best.

Same.

**Comment by Quelklef** [comment]
If we're going with this, I think it's important to address one thing:

Clicking between the braces to reveal the pragmas signifies (to me at least) that one should click between the braces again to hide the pragmas. So the shown content should be clickable rather than the braces.

It would be a little weird, though, and frankly I don’t think doing it this way is a great idea. The content shown is conceptually that--content, not navigation. However, I'm not sure a great alternative. Show content on hover-on and hide it in hover-off would work, but that's also a bit janky.

**Comment by Araq** [comment]
> If we're going with this, I think it's important to address one thing:

Clicking between the braces to reveal the pragmas signifies (to me at least) that one should click between the braces again to hide the pragmas. So the shown content should be clickable rather than the braces.

Let's just use ``{.+.}`` then and expanded ``{.- noSideEffect, more, stuff.}`` or something along these lines to give us enough visual clues.

**Comment by dom96** [comment]
Nah, there is little need to hide the pragma after it has been shown. I say just keep the pragmas there with no option to hide them.

**Comment by Araq** [comment]
Or that, ok.

**Comment by Quelklef** [comment]
OK to change docgen.nim for this, or should be pure JS/CSS?

**Comment by dom96** [comment]
We don't use JS when we can help it.

**Comment by Quelklef** [comment]
Unfortunately, because CSS is nondeterministic, I couldn't style it quite as nicely as I'd have liked, but here is the current state:
screenshot from 2018-07-03 17-28-28

File: config/nimdoc.cfg [hunk 1]
```diff
@@ -1330,15 +1330,23 @@ dt pre > span.Operator ~ span.Identifier {
...
   margin-bottom: -5px; }
+<<<<<<< HEAD
```
Comment (kaushalmodi) [inline config/nimdoc.cfg]: You committed some merge conflicts by mistake.

File: config/nimdoc.cfg [hunk 2]
```diff
@@ -1330,15 +1330,23 @@ dt pre > span.Operator ~ span.Identifier {
...
+}
+span.pragmabegin {
+  border-right: none;
 }
 span.pragmaend {
-  cursor: pointer;
+  border-left: none;
 }
+=======
+>>>>>>> Visual cues at hidden pragmas in docs.
```
Comment (kaushalmodi) [inline config/nimdoc.cfg]: See those "<<<<<<<" and ">>>>>>>".

File: config/nimdoc.cfg (hunk 1 above)
Comment (Quelklef) [reply config/nimdoc.cfg]: Ah, dammit. My bad. Will fix now.

File: compiler/docgen.nim [hunk 3]
```diff
@@ -222,6 +222,15 @@ proc getPlainDocstring(n: PNode): string =
       result = getPlainDocstring(n.sons[i])
       if result.len > 0: return
 
+func htmlCollapse(html: string): string =
+    ## Remove leading whitespace and trailing newlines from every line
+    result = ""
+    for line in html.split("\n"):
+        for i, c in line:
```
Comment (dom96) [inline compiler/docgen.nim]: Two space indent

Comment (dom96) [reply compiler/docgen.nim]: Use ``splitLines`` and ``strip``.

File: compiler/docgen.nim [hunk 4]
```diff
@@ -222,6 +222,10 @@ proc getPlainDocstring(n: PNode): string =
       result = getPlainDocstring(n.sons[i])
       if result.len > 0: return
 
+func htmlCollapse(html: string): string =
```
Comment (Araq) [inline compiler/docgen.nim]: Don't use ``func``, it's too new. Also it's used nowhere else in ``docgen.nim``.

File: compiler/docgen.nim [hunk 5]
```diff
@@ -222,6 +222,10 @@ proc getPlainDocstring(n: PNode): string =
...
+  ## Strip whitespace and remove trailing newlines from every line
+  return html.splitLines().mapIt(strip(it)).join("")
```
Comment (Araq) [inline compiler/docgen.nim]: Why is this required anyway?

Comment (Araq) [reply compiler/docgen.nim]: This implementation has terrible performance. ;-)

Comment (Quelklef) [reply compiler/docgen.nim]: I wrote it so I could format the HTML later in the code. Since that HTML ends up inside of a `<pre>`, it has to be collapsed.

The poor performance is from the string concatenation, yes? I should be using `result.add`?

Comment (Quelklef) [reply compiler/docgen.nim]: Actually, a [quick benchmark](https://gist.github.com/Quelklef/1f5a7a4c5208ac6ba2569aac4e9ffde2) on my machine shows a ~1.6x speedup using `result.add`, except when in `-d:release`, in which case it's about the same or slower! So I'm not totally sure what optimization you're looking for :P

**Comment by Quelklef** [comment]
Ah, something that I thought was fixed is acting up again. Don't merge yet.

**Comment by Quelklef** [comment]
It should be good now. Unfortunately, I only have Chrome on my dev laptop so I have only tested on Chrome. I figure the changes are simple enough that it shouldn't be a problem. I can rig something up to do cross-browser testing if you'd like, though,

**Comment by Quelklef** [comment]
Update: apparently I also have Firefox 58.0 installed. It looks fine.
